main/src/ca/uwaterloo/flix/language/phase/Optimizer.scala

```diff
@@ -41,14 +40,263 @@ object Optimizer extends Phase[SimplifiedAst.Root, SimplifiedAst.Root] {
+          case Expression.False => (optimizeExp(exp3)._1, true)
+          case _ =>
+            val (e2, b2) = optimizeExp(exp2)
+            val (e3, b3) = optimizeExp(exp3)
+            (Expression.IfThenElse(e1, e2, e3, tpe, loc), b1 || b2 || b3)
+        }
+      case Expression.Let(sym, exp1, exp2, tpe, loc) =>
+        val (e1, b1) = optimizeExp(exp1)
+        val (e2, b2) = optimizeExp(exp2)
+        (Expression.Let(sym, e1, e2, tpe, loc), b1 || b2)
+      case Expression.Is(exp, tag, loc) =>
+        val (e, b) = optimizeExp(exp)
+        (Expression.Is(e, tag, loc), b)
+      case Expression.Tag(sym, tag, exp, tpe, loc) =>
+        val (e, b) = optimizeExp(exp)
+        (Expression.Tag(sym, tag, e, tpe, loc), b)
+      case Expression.Untag(tag, exp, tpe, loc) =>
+        val (e, b) = optimizeExp(exp)
+        (Expression.Untag(tag, e, tpe, loc), b)
+      case Expression.Index(base, offset, tpe, loc) =>
+        val (e, b) = optimizeExp(base)
+        (Expression.Index(e, offset, tpe, loc), b)
+      case Expression.Tuple(elms, tpe, loc) =>
+        val (es, bs) = optimizeExps(elms).unzip
+        (Expression.Tuple(es, tpe, loc), bs.exists(identity))
+      case Expression.Existential(fparam, exp, loc) =>
+        val (e, b) = optimizeExp(exp)
+        (Expression.Existential(fparam, e, loc), b)
+      case Expression.Universal(fparam, exp, loc) =>
+        val (e, b) = optimizeExp(exp)
+        (Expression.Universal(fparam, e, loc), b)
+      case Expression.NativeConstructor(constructor, args, tpe, loc) =>
+        val (es, bs) = optimizeExps(args).unzip
+        (Expression.NativeConstructor(constructor, es, tpe, loc), bs.exists(identity))
+      case Expression.NativeField(field, tpe, loc) => (Expression.NativeField(field, tpe, loc), false)
+      case Expression.NativeMethod(method, args, tpe, loc) =>
+        val (es, bs) = optimizeExps(args).unzip
+        (Expression.NativeMethod(method, es, tpe, loc), bs.exists(identity))
+      case Expression.UserError(tpe, loc) => (Expression.UserError(tpe, loc), false)
+      case Expression.MatchError(tpe, loc) => (Expression.MatchError(tpe, loc), false)
+      case Expression.SwitchError(tpe, loc) => (Expression.SwitchError(tpe, loc), false)
     }
 
-    // TODO: Implement.
+    /**
+      * Fully optimizes the given Expression `e0`.
+      *
+      * That is, iteratively applies `optimizeExp` to `e0` until no further optimizations are made.
+      */
+    def fullyOptimizeExp(e0: Expression): Expression = {
+      var exp = e0
+      var changing = true
 
-    root.toSuccess
-  }
+      while (changing) {
+        val (optExp, change) = optimizeExp(exp)
+        exp = optExp
+        changing = change
+      }
+
+      exp
+    }
 
+    /**
+      * Fully optimizes each SimplifiedAst.Term.Head in `hs`.
+      */
+    def fullyOptimizeHeadTerms(hs: List[SimplifiedAst.Term.Head]): List[SimplifiedAst.Term.Head] = hs.map(fullyOptimizeHeadTerm)
```

Comment:
I think there is no reason to optimize constraints; they can only contain literals or applications of functions with literals (as far as I recall)